Support narrowing literals and enums using the in operator in combination with list, set, and tuple expressions.

tyralla opened the pull request:
Note: This PR initially dealt only with tuple expressions. 

---

The general idea is to transform expressions like

```python
(x is None) and (x in (1, 2)) and (x not in (3, 4))
```

into

```python
(x is None) and (x == 1 or x == 2) and (x != 3 and x != 4)
```

This transformation circumvents the need to extend the (already complicated) narrowing logic further.

See the test `testNarrowTypeAfterInTuple`, which shows a slight change in the narrowing behaviour.  I don't know if this is an improvement, but at least it seems to increase consistency.

The only downside of the transformation approach I am aware of is that it might cause some hard-too-understand error messages.   The only such problem I encountered is the `Right operand of "and" is never evaluated` warning in the `testNarrowLiteralsNotInTupleExpression` test.

tyralla committed:
Support narrowing literals and enums using the in operator in combination with tuple expressions.

The general idea is to transform expressions like

(x is None) and (x in (1, 2)) and (x not in (3, 4))

into

(x is None) and (x == 1 or x == 2) and (x != 3 and x != 4)

This transformation circumvents the need to extend the (already complicated) narrowing logic further.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tyralla committed:
fix

tyralla committed:
Merge branch 'feature/narrow_using_in' of https://github.com/tyralla/mypy into feature/narrow_using_in

tyralla committed:
Optional[Expression] -> Expression | None

tyralla committed:
Add __eq__ to object in tuple.pyi

davidhalter commented:
It feels a bit weird that this is the way we approach this. I have recently done a lot of research into the Mypy narrowing code and it feels like the most convoluted and complicated part of the whole Mypy code base. (Other parts like the recently added --new-type-inference are way more approachable even though they are more complex). I therefore understand that you wanted to generate special case this.

However, in this case it feels like the code would be more approachable if it was actually implemented in the `in` operator narrowing part of the code. It just feels super weird that there would be two places that narrow None in different ways.

Also you handled `TupleExpr`, which would imply a different behavior for `TupleType`, but that's fixable, I guess.

tyralla committed:
simplification: avoid returning None

tyralla commented:
@davidhalter Thanks for your opinion on this, which I understand.  I started with extending the mentioned `in` operator section and targeting `TupleType` but felt that a separate handling for `TupleExpr` would be simpler and maybe more correct.

Regarding the different ways to narrow `None`, I think this problem already exists.  This is how the `testNarrowTypeAfterInTuple` test case looks like without my changes:

```python
from typing import Optional
class A: pass
class B(A): pass
class C(A): pass

y: Optional[B]
if y in (B(), C()):
    reveal_type(y) # N: Revealed type is "__main__.B"
else:
    reveal_type(y) # N: Revealed type is "Union[__main__.B, None]"
```

But if one uses `==` instead of `in`, the result (again, without my changes) is:

```python
if y == B() or y == C():
    reveal_type(y) # N: Revealed type is "Union[__main__.B, None]"
else:
    reveal_type(y) # N: Revealed type is "Union[__main__.B, None]"
```

So, for this aspect, I would say the proposed change increases consistency (always the second result).

I might have been too cautious about the `TupleExpr` vs. `TupleType` issue.  I wanted to avoid handling code where users apply `tuple` subclasses with overwritten `__contains__` methods.  However, I now checked it and realised that Mypy models such subclasses with `Instance` and not `TupleType`.  So, generally, targeting `TupleType` seems the right choice (although I expect 99 % of all potential use cases to work with tuples created on the fly so that targeting `TupleExpr` only could be satisfying enough).

Two Mypy primer responses are satisfying (discord), and one needs investigation (mkosi).  However, for Mypyc, there seems to be a relevant problem.

tyralla committed:
replace the critical `in` comparisons for testing

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tyralla committed:
Revert "[pre-commit.ci] auto fixes from pre-commit.com hooks"

This reverts commit dea2614f4f7023280884e0366a82d81a85e92481.

tyralla committed:
Revert "replace the critical `in` comparisons for testing"

This reverts commit 207c56eb3cd7732fbaa82e2f7e10736bf127a608.

tyralla committed:
fix mypyc crash

tyralla committed:
make NameExpr mypyc copyable

tyralla committed:
ignore star expressions

tyralla committed:
update docs

tyralla commented:
The mkosi and mypyc problems are fixed.  I also included a small documentation update in case this pull request gets accepted.  It's clearly not the most general solution, but it should handle the most frequent use case (at least the one often relevant to me...), so I will wait for a review.

tyralla committed:
Also support list expressions.

tyralla committed:
Merge branch 'master' into feature/narrow_using_in

# Conflicts:
#	test-data/unit/check-narrowing.test

tyralla commented:
@JukkaL @Jordandev678 

I expected this PR to be superseded by the more general PR #17344 and wanted to close it.  However, when trying out the new feature in Mypy 3.13, I realised that #17865 reverted #17344.  So, I reconsidered my approach and still find it not too bad for a review.  

If the "transformation approach" is unacceptable, could we combine the ideas of this PR and #17344?

Also, I widened the scope of this PR by supporting list expressions.  Using list expressions for this kind of narrowing is equally fine as using tuple expressions, and by supporting it, I may have an excuse for not supporting tuple types...

I also considered supporting set expressions but was unsure because of missing (plain) literals for empty sets and single-value sets.  On the other hand, who would write `if x in set()` or `if x in set([1])`?

tyralla commented:
The new Mypy primer finding indicates a small flaw in the `xarray` library.  Small repro:

```python
from typing import Literal

PadModeOptions = Literal[
    "constant",
    "edge",
    "linear_ramp",
    "maximum",
    "mean",
    "median",
    "minimum",
    "reflect",
    "symmetric",
    "wrap",
]

def pad(mode: PadModeOptions = "constant") -> None:
    # if mode in ("edge", "reflect", "symmetric", "wrap"):  # actual code, Mypy only reports an error with this PR included
    if mode == "edge" or mode == "reflect" or mode == "symmetric" or mode == "wrap"  # for this similar code, Mypy would also report an error without this PR
        coord_pad_mode = mode
    else:
        coord_pad_mode = "constant"
```

tyralla committed:
Also support set expressions.

tyralla commented:
> I also considered supporting set expressions but was unsure because of a missing literal for empty sets. On the other hand, who would write `if x in set()`?

After rethinking, there is really no good reason not to support set expressions, so I extended the PR's title again...

tyralla committed:
Merge branch 'master' into feature/narrow_using_in

# Conflicts:
#	mypy/checker.py
#	mypyc/test-data/run-tuples.test
#	test-data/unit/check-narrowing.test
#	test-data/unit/fixtures/tuple.pyi

tyralla committed:
Adjust the `testNarrowingOptionalEqualsNone` test case.

A recent change that I just merged seems to interact positively with this PR: increase in consistency when narrowing away `None`.

tyralla committed:
update comment

hauntsaninja commented:
I've fixed this to some extent in 1.20 and will fix the remainder soon
